src/modules/article/article.service.ts

```diff
@@ -7,75 +7,104 @@ import { MessageCodeError } from '../../shared/errors/message-code-error';
+      if (!category) throw new MessageCodeError('category:notFound');
+      categories.push(category);
+    }
 
     const newArticle = this.articleRepository.create(data);
-    await this.articleRepository.save(data);
-    return newArticle;
+    newArticle.seoId = seoId;
+    newArticle.categories = categories;
+    return this.articleRepository.save(newArticle);
   }
 
   async update(data: UpdateArticleDto): Promise<Article> {
     const article = await this.articleRepository.findOne(data.id);
     if (!article) throw new MessageCodeError('article:notFound');
+    if (data.title) {
+      const seoId = ToTranslit(data.title);
+      const articleExist = !!await this.articleRepository.findOne({ seoId });
+      if (articleExist) throw new MessageCodeError('article:exist');
+      article.seoId = seoId;
+    }
+    if (data.categoriesIds) {
+      let categoryId;
+      for (categoryId of data.categoriesIds) {
+        const existCategory = await this.categoryRepository.findOne({ id: categoryId });
+        if (!existCategory) throw new MessageCodeError('category:notFound');
+      }
+    }
     Object.assign(article, data);
     await this.articleRepository.save(article);
     return article;
   }
 
   async remove(id: string): Promise<boolean> {
-    const deleteResponse = await this.articleRepository.delete(id);
+    const deleteResponse = await this.articleRepository.createQueryBuilder()
+      .delete()
+      .from(Article)
+      .where('id = :id', { id })
+      .execute();
     return !!deleteResponse.affected;
   }
 
   async getArticlesByGender(gender: string): Promise<Article[]> {
-    return this.articleRepository.find({
-      where: { gender },
-    });
+    return this.articleRepository.createQueryBuilder('article')
+      .leftJoinAndSelect('article.categories', 'category')
+      .where('article.gender IN (:...genders)', { genders: [gender, 'female both male'] }).getMany();
   }
 
   async getArticlesByCategoryId(categoryId: string): Promise<Article[]> {
-    return this.articleRepository.find({
-      where: { categoryId },
-    });
+    // #TODO: not displayed all categories to which the article belongs
+    return this.articleRepository.createQueryBuilder('article')
+      .leftJoinAndSelect('article.categories', 'category')
+      .where('category.id = :id', { id: categoryId })
+      .getMany();
   }
 
-  async getArticlesByGenderAndCategory(gender: string, categoryId: string): Promise<Article[]> {
-    return this.articleRepository.find({
-      where: { gender, categoryId },
-    });
+  async getArticlesByGenderAndCategoryId(gender: string, categoryId: string): Promise<Article[]> {
+    return this.articleRepository.createQueryBuilder('article')
+      .leftJoinAndSelect('article.categories', 'category')
+      .where('category.id = :id', { id: categoryId })
+      .andWhere('article.gender IN (:...genders)', { genders: [gender, 'female both male'] })
```

Comment:
'female both male' why didn't you put genders into constants?

Reply:
done